File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/AbstractCountCategorizer.java
package cz.fb.manaus.reactor.categorizer;

import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.model.SettledBet;

import java.util.Collections;
import java.util.Set;

abstract public class AbstractCountCategorizer implements SettledBetCategorizer {

    private final String prefix;
    private final int maxCount;

    protected AbstractCountCategorizer(String prefix, int maxCount) {
        this.prefix = prefix;
        this.maxCount = maxCount;
    }

    @Override
    public boolean isMarketSnapshotRequired() {
        return true;
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        int count = getCount(settledBet);
        return Collections.singleton(prefix + toCategory(count));
    }

    protected abstract int getCount(SettledBet bet);

    private String toCategory(int count) {
        if (count >= maxCount) {
            return maxCount + "+";
        } else {
            return Integer.toString(count);
        }
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/AbstractCountCategorizer.kt
package cz.fb.manaus.reactor.categorizer

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.repository.domain.RealizedBet
import cz.fb.manaus.core.repository.domain.SettledBet

abstract class AbstractCountCategorizer(private val prefix: String, private val maxCount: Int) : RealizedBetCategorizer {

    override val isMarketSnapshotRequired: Boolean = true

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val count = getCount(realizedBet)
        return setOf(prefix + toCategory(count))
    }

    protected abstract fun getCount(bet: SettledBet): Int

    private fun toCategory(count: Int): String {
        return if (count >= maxCount) {
            maxCount.toString() + "+"
        } else {
            Integer.toString(count)
        }
    }
}
